### PR TITLE
Log an error when an integration can't proceed due to auth issues

### PR DESCRIPTION
**Pull request by base10**
* With SyncJob, this traps the exception
* Icims::CandidatesImporter was trapping the exception, now it logs
* Greenhouse::CandidatesImporter does not trap the exception, which
  was the expected behavior previous to notifying the user
  * Greenhouse::CandidatesImporter will now log the error
* Jobvite does not yet have an implementation that would be affected by
  this
* For: https://trello.com/c/I6f1verE/

### CODE REVIEW

**Comment by base10** [comment]
Ideas welcome on some of the exception logging duplication.

File: app/jobs/sync_job.rb [hunk 1]
```diff
@@ -16,14 +19,27 @@ def connection
   end
 
   def deliver_notification(results)
-    SyncMailer.sync_notification(
-      email: user.email,
-      integration_id: @integration_id,
-      results: results
-    ).deliver
+    if !erred?
```
Comment (jferris) [inline app/jobs/sync_job.rb]: I don't think the `erred?` logic is necessary, because an exception will skip the `deliver_notification` and jump into the `rescue` clause of `perform`.

File: app/jobs/sync_job.rb [hunk 2]
```diff
@@ -2,11 +2,14 @@ class SyncJob
   def initialize(integration_id, user_id)
     @integration_id = integration_id
     @user_id = user_id
+    @erred = false
   end
 
   def perform
     results = connection.sync
     deliver_notification results
+  rescue NetSuite::Client::Unauthorized => exception
```
Comment (jferris) [inline app/jobs/sync_job.rb]: Since this is a generic class (not specific to NetSuite), do we want to make a generic `Unauthorized` exception and use it here?

File: app/jobs/sync_job.rb [hunk 3]
```diff
@@ -16,14 +19,27 @@ def connection
...
+        results: results
+      ).deliver
+    end
+  end
+
+  def handle_authentication_exception(exception)
+    @erred = true
+    Rails.logger.error(
+      "#{exception.class} error #{exception.message} for user_id: #{user.id}"
+    )
```
Comment (jferris) [inline app/jobs/sync_job.rb]: Do we want to deliver a notification to the user here?

File: app/jobs/sync_job.rb (hunk 1 above)
Comment (base10) [inline app/jobs/sync_job.rb]: I was seeing something different when I was working on this as `NetSuite::Export::Job`, but let me give that a shot.

Comment (base10) [inline app/jobs/sync_job.rb]: Tested and works without `erred?`. Yay. Fixed with 278fb6a

File: app/jobs/sync_job.rb (hunk 3 above)
Comment (base10) [inline app/jobs/sync_job.rb]: A user will get a message when the exception is raised in `NetSuite::Client`

File: app/jobs/sync_job.rb (hunk 2 above)
Comment (base10) [inline app/jobs/sync_job.rb]: I did like having <Integration>::<class>::Unauthorized across the four exceptions since that made it clear where the exception was raised, but I'm open to consolidating them to one exception class.

File: app/jobs/sync_job.rb (hunk 3 above)
Comment (jferris) [inline app/jobs/sync_job.rb]: If we handle it here instead of in the clients, we'll only have to implement it once for NetSuite/Jobvite, right? And future, on-demand integrations will get it "for free?"

Comment (base10) [inline app/jobs/sync_job.rb]: The expectation is that the import controller is going to effectively run a sync job vs. calling the import directly, I'm good with that.

Comment (base10) [inline app/jobs/sync_job.rb]: Addressed with 3fbf492

File: app/jobs/sync_job.rb (hunk 2 above)
Comment (base10) [inline app/jobs/sync_job.rb]: Addressed with 9b5298d. Thoughts on consolidating the other integration exceptions in this PR vs. a separate one / as needed?

**Comment by base10** [comment]
@jferris Feedback addressed, ready for rereview. Have one question for you: https://github.com/namely/connect/pull/134#discussion_r33792524

File: app/jobs/sync_job.rb (hunk 2 above)
Comment (jferris) [inline app/jobs/sync_job.rb]: I think it would be fine to do in a separate pull request. I think it's a good idea, though - every integration will have to handle authorization errors, so we may as well have a single exception that represents the common concept.

**Comment by jferris** [comment]
Ready to squash and merge at your discretion.

**Comment by base10** [comment]
Thank you, @jferris!
